scripts/lib/aws/elasticache.sh

```diff
@@ -316,13 +438,105 @@ k8s_assert_replication_group_replica_count() {
     exit 1
   fi
-  local replication_group_id="$1"
-  local expected_value="$2"
-  local nodeGroupMembersCount="$(kubectl get ReplicationGroup/"$replication_group_id" -o json | jq -r -e '.status .nodeGroups[0] .nodeGroupMembers' | jq length)"
-  actual_replica_count=$(( nodeGroupMembersCount - 1 ))
-  if [[ "$expected_value" -ne "$actual_replica_count" ]]; then
-    echo "FATAL: replication group $replication_group_id Node Groups count on k8s cluster. expected: $expected_value, actual: $actual_replica_count"
+  local node_group_size=$(k8s_get_rg_field "$1" ".status .nodeGroups[0] .nodeGroupMembers" | jq length)
+  actual_replica_count=$(( node_group_size - 1 ))
+  if [[ "$2" -ne "$actual_replica_count" ]]; then
+    echo "FAIL: expected $2 replicas per node group for replication group $1, actual: $actual_replica_count"
     print_k8s_ack_controller_pod_logs
     exit 1
   fi
 }
+
+# k8s_assert_replication_group_total_node_count asserts the total number of nodes/clusters in the specified
+#   replication group. The total node count should be (# shards) x (# replicas per shard + 1)
+# k8s_assert_replication_group_total_node_count requires 2 arguments
+#     replication_group_id
+#     expected_count - expected total node count
+k8s_assert_replication_group_total_node_count() {
+  if [[ $# -ne 2 ]]; then
+    echo "FATAL: Wrong number of arguments passed to ${FUNCNAME[0]}"
+    echo "Usage: ${FUNCNAME[0]} replication_group_id expected_count"
+    exit 1
+  fi
+  local actual_value=$(k8s_get_rg_field "$1" ".status .memberClusters" | jq length)
+  if [[ "$2" != "$actual_value" ]]; then
+    echo "FAIL: expected $2 total nodes for replication group $1, actual: $actual_value"
+    print_k8s_ack_controller_pod_logs
+    exit 1
+  fi
+}
+
+# delete all replication groups in existing clusters and print a debug message
+k8s_perform_rg_test_cleanup () {
+  debug_msg "Cleaning up test replication groups..."
+  kubectl delete ReplicationGroup --all 2>/dev/null
+  assert_equal "0" "$?" "Expected success from kubectl delete but got $?" || exit 1
+}
+
+# assert_terminal_condition_true asserts that the terminal condition exists, has status "True", and
+#   the message associated with the terminal condition matches the one provided. The assertions should pass
+#   following an invalid operation (e.g. create RG with negative replica count, e.g. modify cluster mode
+#   disabled RG from 1 to 2 shards)
+# assert_terminal_condition_true requires 2 arguments
+#   replication_group_id
+#   expected_substring: a substring of the expected message associated with the terminal condition
+assert_rg_terminal_condition_true() {
+  if [[ $# -ne 2 ]]; then
+    echo "FATAL: Wrong number of arguments passed to ${FUNCNAME[0]}"
+    echo "Usage: ${FUNCNAME[0]} replication_group_id expected_substring"
+    exit 1
+  fi
+
+  terminal_cond=$(k8s_get_rg_field "$1" ".status .conditions[]" | jq -r -e 'select(.type == "ACK.Terminal")')
```

Comment:
I believe this should be `".status.conditions[]"`, no?

Reply:
I'll be going through and fixing spaces in jq statements